App/Redux/CameraRollRedux.ts

```diff
@@ -85,6 +109,95 @@ export function reducer (state: CameraRollState = initialState, action: CameraRo
+        } else {
+          // add the image to the thread's pending shares
+          threadShares[action.payload.threadId].push(image)
+        }
+      }
+      return {...state, pendingShares: threadShares}
+    case getType(actions.addComment):
+      if (state.pendingShares[action.payload.threadId] === undefined) {
+        return state
+      }
+      const existing = state.pendingShares[action.payload.threadId].find(img=>img.origURL === action.payload.image.origURL)
+      if (!existing || existing.caption !== undefined) {
+        return state
+      }
+      existing.caption = action.payload.caption
+
+      const thread = state.pendingShares[action.payload.threadId].filter((img) => img.origURL !== action.payload.image.origURL)
+      thread.push(existing)
+
+      return {
+        ...state,
+        pendingShares: {
+          ...state.pendingShares,
+          [action.payload.threadId]: thread
+        }
+      }
+    case getType(actions.localPinSuccess):
+      if (state.pendingShares[action.payload.threadId] === undefined) {
+        return state
+      }
+      const partial = state.pendingShares[action.payload.threadId].find(img=>img.origURL === action.payload.image.origURL)
+      if (!partial || partial.addResult !== undefined) {
+        return state
+      }
+      partial.addResult = action.payload.addResult
+
+      const ps = state.pendingShares[action.payload.threadId].filter((img) => img.origURL !== action.payload.image.origURL)
+      ps.push(partial)
+
+      return {
+        ...state,
+        pendingShares: {
+          ...state.pendingShares,
+          [action.payload.threadId]: ps
+        }}
+    case getType(actions.remotePinSuccess):
+    case getType(actions.imagePinError):
+      if (state.pendingShares[action.payload.threadId] === undefined) {
+        return state
+      }
+
+      const filteredShares = state.pendingShares
+      const newArray = filteredShares[action.payload.threadId].filter((img) => img.origURL !== action.payload.image.origURL)
+      if (newArray.length > 0) {
+        filteredShares[action.payload.threadId] = newArray
+      } else {
+        // remove the key from the object if array=0
+        delete filteredShares[action.payload.threadId]
+      }
+      return {...state, pendingShares: filteredShares}
+    case getType(actions.cancelShare):
+      if (state.pendingShares[action.payload.threadId] === undefined) {
+        return state
+      }
+      const filteredArray = state.pendingShares[action.payload.threadId].filter((img) => img.origURL !== action.payload.image.origURL)
+      if (filteredArray.length < 1) {
+        const pendingShares = state.pendingShares
+        delete pendingShares[action.payload.threadId]
+        return {...state, pendingShares}
+      } else {
+        return {
+          ...state,
+          pendingShares: {
+            ...state.pendingShares,
+            [action.payload.threadId]: filteredArray
+          }
+        }
+      }
```

Comment:
Lots of new logic in here, hard to follow. Would be good to update the tests ;)

Reply:
adding a few now